### PR TITLE
feat/keep namespace filter

### PR DESCRIPTION
**Pull request by mosheavni**
Save namespace filter to the `kubectl.json` file
also more elegant approach to saving a state to a file. we can reuse in the future.

Edit:
well.. sometimes it works, and sometimes it doesn't, 
when it works it flashes between namespaced -> all -> namespaced
when it doesn't, it flashes  namespaced -> all

which could cause epilepsy:

### CODE REVIEW

**Comment by mosheavni** [comment]
Fixed! NS filter is working :)

File: lua/kubectl/actions/commands.lua [hunk 1]
```diff
@@ -146,14 +147,16 @@ function M.load_config(file_name)
   if ok then
     return decoded
   end
-  return nil
+  return { session = { view = "pods", namespace = "All" }, filter_history = {} }
```
Comment (Ramilito) [inline lua/kubectl/actions/commands.lua:150]: This adds a connection to the content/model of the config file, I'd rather have the state/session_manager own that.
The idea is to follow the SOLID principle (The S in solid, separation of responsibility), so we should return nil here and let the state/session_manager decide what that means

File: lua/kubectl/actions/commands.lua [hunk 2]
```diff
@@ -146,14 +147,16 @@ function M.load_config(file_name)
...
 end
 
 --- Save to config file
 --- @param file_name string The filename to save
 --- @param data table The content to save
 function M.save_config(file_name, data)
-  local ok, encoded = pcall(vim.json.encode, data)
+  local config = M.load_config("kubectl.json") or {}
+  local merged = vim.tbl_deep_extend("force", config, data)
```
Comment (Ramilito) [inline lua/kubectl/actions/commands.lua:158]: yeah I was too lazy to handle table merging 😄, this looks great!

File: lua/kubectl/actions/commands.lua (hunk 1 above)
Comment (mosheavni) [reply lua/kubectl/actions/commands.lua]: ok returning `nil` works without changing anything

File: lua/kubectl/views/namespace/init.lua [hunk 3]
```diff
@@ -35,10 +35,9 @@ function M.View()
 end
 
 function M.changeNamespace(name)
-  if name == "All" then
-    state.ns = "All"
-  else
-    state.ns = name
+  state.ns = name
+  commands.save_config("kubectl.json", { session = { namespace = name } })
+  if name ~= "All" then
```
[the remark this answers is not in the document]
Comment (mosheavni) [reply lua/kubectl/views/namespace/init.lua:40]: NP moving to `M.close()`

File: lua/kubectl/actions/commands.lua [hunk 4]
```diff
@@ -166,15 +169,16 @@ function M.save_config(file_name, data)
 end
 
 function M.restore_session()
-  local session = M.load_config("kubectl.session.json")
```
[the remark this answers is not in the document]
Comment (mosheavni) [reply lua/kubectl/actions/commands.lua:169]: i'm not sure I understand, can you elaborate? or add to this PR?

File: lua/kubectl/views/filter/init.lua [hunk 5]
```diff
@@ -6,12 +6,10 @@ local tables = require("kubectl.utils.tables")
...
 
 local function save_history(input)
-  local history = commands.load_config("kubectl_filter_history.json")
+  local config = commands.load_config("kubectl.json")
   local history_size = 5
 
-  if history == nil then
-    history = {}
-  end
+  local history = config and config.filter_history and config.filter_history or {}
```
Comment (mosheavni) [inline lua/kubectl/views/filter/init.lua:12]:
```suggestion
  local history = config and config.filter_history or {}
```

File: lua/kubectl/init.lua [hunk 6]
```diff
@@ -122,7 +122,7 @@ vim.api.nvim_create_autocmd("VimLeavePre", {
   callback = function()
     local ok, buf_name = pcall(vim.api.nvim_buf_get_var, 0, "buf_name")
     if ok then
-      commands.save_config("kubectl.session.json", { view = buf_name })
+      commands.save_config("kubectl.json", { session = { view = buf_name } })
```
Comment (mosheavni) [inline lua/kubectl/init.lua:125]:
```suggestion
      commands.save_config("kubectl.json", { session = { view = buf_name, namespace = state.ns } })
```

File: lua/kubectl/actions/commands.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ramilito) [reply lua/kubectl/actions/commands.lua:169]: I'll elaborate and add to this PR!
What I'm trying to push for is sort of having the state management inside the state.lua file.
The commands file only saves and loads the file and shouldn't really know any of the intrnal workings of the session logic.
The same would be in the filter view, it should ask the state.lua module after list of history and shouldn't need to know about where that list comes from.
The same for namespace, it should also just ask the state.lua for the current namespace.

I know that I've taken shortcuts and it's ok if you want to push this in and create a pr for a refactor

[the remark this answers is not in the document]
Comment (mosheavni) [reply lua/kubectl/actions/commands.lua:169]: @Ramilito  i do want to push it and create a different PR for moving to `state.lua`

[the remark this answers is not in the document]
Comment (Ramilito) [reply lua/kubectl/actions/commands.lua:169]: lets do it!

**Comment by Ramilito** [comment]
Did a small refactor if you wanted to look at what I ment!

**Pull request merged**